### PR TITLE
Clean up GPU memory after killing sglang processes

### PR DESCRIPTION
**Pull request by MrAta**
## Motivation
Clean up GPU memory after killing sglang processes

## Modifications
Modify the killall_sglang.sh script to clean up any fore and background process using up the GPU memory.

Currently, running the script doesn't clean up the memory properly (see screenshot below), with this PR it should be fixed.
Screenshot 2024-12-11 at 7 27 42 PM

## Checklist

- [x] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [ ] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [ ] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

File: scripts/killall_sglang.sh [hunk 1]
```diff
@@ -8,7 +8,6 @@ kill -9 $(ps aux | grep 'sglang::' | grep -v 'grep' | awk '{print $2}') 2>/dev/n
 kill -9 $(ps aux | grep 'sglang.launch_server' | grep -v 'grep' | awk '{print $2}') 2>/dev/null
 kill -9 $(ps aux | grep 'sglang.bench' | grep -v 'grep' | awk '{print $2}') 2>/dev/null
 
-# Clean all GPU processes if any argument is provided
-if [ $# -gt 0 ]; then
-    kill -9 $(nvidia-smi | sed -n '/Processes:/,$p' | grep "   [0-9]" | awk '{print $5}') 2>/dev/null
-fi
+# Clean all GPU processes
+kill -9 $(nvidia-smi | sed -n '/Processes:/,$p' | grep "   [0-9]" | awk '{print $5}') 2>/dev/null
+lsof /dev/nvidia* | awk '{print $2}' | xargs kill -9 2>/dev/null
```
Comment (ByronHsu) [inline scripts/killall_sglang.sh]: can we show nvidia-smi again at the end to ensure everything is cleaned up? seems that 
image
still has issues

Comment (MrAta) [reply scripts/killall_sglang.sh]: this one is from another PR which doesn't contain this fix tho. updated the script to show gpu status again.

Comment (ByronHsu) [reply scripts/killall_sglang.sh]: ata is fast :rocket:

**Comment by ByronHsu** [comment]
@zhyncs

File: scripts/killall_sglang.sh [hunk 2]
```diff
@@ -8,7 +8,9 @@ kill -9 $(ps aux | grep 'sglang::' | grep -v 'grep' | awk '{print $2}') 2>/dev/n
 kill -9 $(ps aux | grep 'sglang.launch_server' | grep -v 'grep' | awk '{print $2}') 2>/dev/null
 kill -9 $(ps aux | grep 'sglang.bench' | grep -v 'grep' | awk '{print $2}') 2>/dev/null
 
-# Clean all GPU processes if any argument is provided
-if [ $# -gt 0 ]; then
-    kill -9 $(nvidia-smi | sed -n '/Processes:/,$p' | grep "   [0-9]" | awk '{print $5}') 2>/dev/null
-fi
+# Clean all GPU processes
+kill -9 $(nvidia-smi | sed -n '/Processes:/,$p' | grep "   [0-9]" | awk '{print $5}') 2>/dev/null
```
Comment (zhyncs) [inline scripts/killall_sglang.sh]: I didn't enable it by default because, in some environments, we shouldn't terminate other processes.

Comment (MrAta) [reply scripts/killall_sglang.sh]: >some environment

Like where?

Comment (zhyncs) [reply scripts/killall_sglang.sh]: For example, in our development environment, we share an H100 node, and others might use the GPU to run other tasks.

Comment (zhyncs) [reply scripts/killall_sglang.sh]: This is why I didn't set it as default, but it's also easy to enable when we want to, just by adding an extra parameter.

Comment (MrAta) [reply scripts/killall_sglang.sh]: Hmm OK, lemme update it then.

Comment (MrAta) [reply scripts/killall_sglang.sh]: @zhyncs put back the if condition and enabled killing for the router tests.

**Review by ByronHsu — APPROVED** [review]

**Comment by ByronHsu** [comment]
@MrAta can you fix the conflict

**Comment by MrAta** [comment]
> @MrAta can you fix the conflict

Fixed!

File: .github/workflows/pr-test-rust.yml [hunk 3]
```diff
@@ -60,7 +60,7 @@ jobs:
           pip install --force-reinstall dist/*.whl
       - name: Run e2e test
         run: |
-          bash scripts/killall_sglang.sh
+          bash scripts/killall_sglang.sh "nuk_gpus"
```
Comment (zhyncs) [inline .github/workflows/pr-test-rust.yml:63]: Why does pr-test-rust need this parameter?

Comment (MrAta) [reply .github/workflows/pr-test-rust.yml:63]: Seems like most of the time GPU memory isn't cleaned up properly during those tests; so it's to make sure we can load the model for tests.

**Pull request merged**